Updated SSHD Password Generation

NoMoreFood opened the pull request:
- Updated SSHD user password generation routine to be longer and more complex.  This should satisfy systems with password filters that require more character types or very long passwords.
- Updated routine to now securely zero memory for the SSHD  account password.
- Corrected attempt to write to NULL pointer by localtime_s() in localtime_r() and made function return NULL on error per specification.
- Addressed various compiler / code analysis warnings.

NoMoreFood committed:
Updated SSHD Password Generation

- Updated SSHD user password generation routine to be longer and more complex.  This should satisfy systems with password filters that require more character types or very long passwords.
- Updated routine to now securely zero memory for the temporary password.

NoMoreFood committed:
Addressed Various Compiler Warnings

NoMoreFood committed:
Corrected Implementation of localtime_r()

- Corrected attempt to write to NULL pointer by localtime_s() in localtime_r() and made function return NULL on error per specification.

manojampalam commented:
GUID  is 36 chars long, isn't this long enough?
How is the new logic guaranteed to satisfy systems with complex password requirements?

NoMoreFood commented:
The length is probably fine; it was more the character types.  On my systems in my day job, we require four different character types (upper, lower, number, symbol).  You're right that it doesn't guarantee it in the logical sense, but it would be statistically extremely unlikely to not have one each of those those types given PWLEN is 256.  My back-of-the-hand calculations put this somewhere around 0.000000009% likelihood of not getting a number (the "rarest" type).  We also have a filter for more than two contiguous repeated characters which is still somewhat statistically possible to get when using a GUID.

manojampalam commented:
How about we keep it simple, generate a guid and replace the first 4 chars with Aa%1 ? sshd is not a critical account. GUID is sufficiently complex for our needs.

NoMoreFood commented:
I wasn't concerned so much about the security aspect so much as the ability to pass supplemental DLL password filters and having the account creation fail.  I've seen ones that detect patterns, check for 3+ repeated characters, check for 3+ consecutive characters, and additional character types.  For GUIDs, there's about an 8% chance you'll encounter repeated/consecutive characters given the character set is so limited (hex).  That said, there's only one potential reported issue of it in the issue tracker list plus my own occurrence so it's not super critical.  Your proposed alternative would address the character type check 100% of the time, but we'd be still left with the 8% repeated character chance.

manojampalam commented:
I'm still concerned about the perf impact from the new generator. Crypto calls and strcat calls are both expensive. These are currently run on each sshd service startup. We should definitely add conditional logic to only invoke this, if sshd account is not already present. 

But otherwise, I think if we can plant in Aa%1, we should be good on most environments. Beyond that, there is a very easy workaround - i.e to create sshd account explicitly with whatever pass requirements that may exist.

NoMoreFood commented:
Alright, let me revise it a bit and we can reevaluate.  It only takes a few milliseconds to run on my system, but you're right that we probably need not run it on every startup if the account already exists.

manojampalam commented on contrib/win32/win32compat/wmain_sshd.c:
```diff
@@ -110,33 +110,43 @@ generate_host_keys()
+				password[i] = (password[i] % ((L'~' + 1) - L'!')) + L'!';
+
+			/* add user to local accounts */
+			if (NetUserAdd(NULL, 1, (LPBYTE)&ui, &dwError) != NERR_Success) {
```
Suggest removing the failure path (just ignore the failure). Otherwise, if the failure manifests, sshd service would bail out with no clue to user. If we ignore, sshd will still bail out at a later time, logging an event that sshd account is not found.

NoMoreFood replied:
As you wish.  Just re-committed.

NoMoreFood committed:
Updated SSHD Password Generation - Revised

- Addressed potential performance concerns for SSHD account creation.
- Now only creating account if it does not already exist. This will eliminate unnecessary auditable events generated by attempting to create the account on each service startup.
